File: 190.reverse-bits.cpp
/*
 * @lc app=leetcode id=190 lang=cpp
 *
 * [190] Reverse Bits
 */

// @lc code=start
class Solution {
public:
    // T: O(s), s := no. of bits
    // S: O(1)
    uint32_t reverseBits(uint32_t n) {
        // if (!n) return n;
        int ans = 0;
        for (int i = 0; i < 32; ++i) {
            ans <<= 1;
            ans |= (n & 1);
            n >>= 1;
        }
        return ans;
    }
};
// @lc code=end

File: 190.reverse-bits.py
#
# @lc app=leetcode id=190 lang=python3
#
# [190] Reverse Bits
#

# @lc code=start
class Solution:
    # T: O(s), s := no. of bits
    # S: O(s), s := no. of bits
    def reverseBits(self, n: int) -> int:
        ans = 0
        for _ in range(32):
            ans <<= 1  # shift left 1
            ans |= (n & 1)  # append rightmost bit to ans
            n >>= 1  # remove rightmost bit
        return ans
        
# @lc code=end

